### PR TITLE
test: expand AfriqueGemma integration tests with full language coverage

### PR DESCRIPTION
**Pull request by tobi-legan**
## 🎯 What problem does this PR solve?

- AfriqueGemma translation model had minimal test coverage — only 4 language pairs, no edge cases

### Expanded Translation Tests (`afriquegemma-translation.test.js`)
- 9 core EN→African pairs: Swahili, Yoruba, Hausa, Zulu, Igbo, Amharic, Xhosa, Shona, Kinyarwanda
- Reverse pairs (African→EN)
- Bridge language pairs: French, Portuguese, Arabic (to English and direct to African)
- African-to-African pivot translations (Swahili→English→Yoruba)
- Longer content, mixed content, and sequential calls on same instance

### New Edge Cases Test (`afriquegemma-edge-cases.test.js`)
- Empty/whitespace input handling
- Model lifecycle (load/unload/reload)
- Cancel mid-translation
- Tools-required config validation

### Mobile Runner
- Registered `runAfriquegemmaEdgeCasesTest` in `integration.auto.cjs`

## 🧪 How was it tested?

- Passing on desktop: Linux x64, Linux arm64, Darwin arm64, Darwin x64
- Mobile: registered for Device Farm runs (iOS + Android)

### CODE REVIEW

**Review by olyasir — DISMISSED** [review]

File: packages/qvac-lib-infer-llamacpp-llm/test/integration/finetuning-pause-resume.test.js
Comment (gianni-cor) [inline packages/qvac-lib-infer-llamacpp-llm/test/integration/finetuning-pause-resume.test.js:1]: we need to run finetuning on mobile. please remove this change

**Review by gianni-cor — DISMISSED** [review]

File: .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml [hunk 1]
```diff
@@ -918,324 +929,411 @@ jobs:
...
+            aws devicefarm list-devices --region us-west-2 \
+              --filters '[{"attribute":"MANUFACTURER","operator":"EQUALS","values":["Samsung"]},{"attribute":"MODEL","operator":"CONTAINS","values":["S25 Ultra"]},{"attribute":"PLATFORM","operator":"EQUALS","values":["ANDROID"]}]' \
+              --query 'devices[].{name:name,model:model,os:os,availability:availability}' --output table || true
+
+            echo "Querying available Pixel 9 devices..."
+            aws devicefarm list-devices --region us-west-2 \
+              --filters '[{"attribute":"MANUFACTURER","operator":"EQUALS","values":["Google"]},{"attribute":"MODEL","operator":"CONTAINS","values":["Pixel 9"]},{"attribute":"PLATFORM","operator":"EQUALS","values":["ANDROID"]}]' \
+              --query 'devices[].{name:name,model:model,os:os,availability:availability}' --output table || true
+
+            SAMSUNG_FILTER='{"filters":[{"attribute":"MANUFACTURER","operator":"EQUALS","values":["Samsung"]},{"attribute":"MODEL","operator":"CONTAINS","values":["S25 Ultra"]},{"attribute":"PLATFORM","operator":"EQUALS","values":["ANDROID"]}],"maxDevices":1}'
```
Comment (gianni-cor) [inline .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml]: **Hardcoded device models will fail silently when unavailable.**

Samsung Galaxy S25 Ultra and Google Pixel 9 are hardcoded here. The `|| true` on the listing queries above (lines 1170, 1175) swallows failures, so if these devices are retired from Device Farm the `schedule-run` call will fail with a cryptic error.

Suggested fix — validate availability before scheduling and warn early:

```bash
# After the device listings, before scheduling:
SAMSUNG_AVAIL=$(aws devicefarm list-devices --region us-west-2 \
  --filters '[{"attribute":"MANUFACTURER","operator":"EQUALS","values":["Samsung"]},{"attribute":"MODEL","operator":"CONTAINS","values":["S25 Ultra"]},{"attribute":"PLATFORM","operator":"EQUALS","values":["ANDROID"]}]' \
  --query 'length(devices[?availability==`AVAILABLE`])' --output text 2>/dev/null || echo "0")
PIXEL_AVAIL=$(aws devicefarm list-devices --region us-west-2 \
  --filters '[{"attribute":"MANUFACTURER","operator":"EQUALS","values":["Google"]},{"attribute":"MODEL","operator":"CONTAINS","values":["Pixel 9"]},{"attribute":"PLATFORM","operator":"EQUALS","values":["ANDROID"]}]' \
  --query 'length(devices[?availability==`AVAILABLE`])' --output text 2>/dev/null || echo "0")

if [ "${SAMSUNG_AVAIL:-0}" -eq 0 ] || [ "${PIXEL_AVAIL:-0}" -eq 0 ]; then
  echo "::warning::Device availability: Samsung=$SAMSUNG_AVAIL, Pixel=$PIXEL_AVAIL. Run may be queued or fail."
fi
```

Same issue exists in the NMTCPP mobile workflow at the corresponding lines.

File: .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml [hunk 2]
```diff
@@ -918,324 +929,411 @@ jobs:
...
+            echo "Generated test spec:"
+            cat testspec.yml
+            SPEC_ARN=$(upload_spec testspec.yml "Android")
+            echo "test_spec_arn=$SPEC_ARN" >> $GITHUB_OUTPUT
+          else
+            # iOS: split tests across 2 devices to prevent resource exhaustion crashes
+            # Group A: lighter tests (afriquegemma, api, bitnet, cache, config, finetuning, generation, image)
+            # Group B: heavier tests (modelLoading, moe, multiInstance, ocr, reasoning, slidingContext, toolCalling, utf8)
+            IOS_GREP_A="runAfriquegemma|runApi|runBitnet|runCache|runConfig|runFinetuning|runGeneration|runImage"
+            IOS_GREP_B="runModel|runMoe|runMulti|runOcr|runReasoning|runSliding|runTool|runUtf8"
```
Comment (gianni-cor) [inline .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml]: **`runDynamicToolsTest` is already silently excluded from iOS mobile runs.**

I checked the function list in `integration.auto.cjs` against these grep patterns. There are 18 functions total, but Groups A + B only cover 17. `runDynamicToolsTest` (from `dynamic-tools.test.js`) starts with `runDynamic` which doesn't match any prefix in either pattern.

More broadly, any new test file added to `test/integration/` will generate a new `run*Test` function in `integration.auto.cjs` but won't appear in these patterns unless someone manually updates them here.

Suggested fix — auto-derive the patterns from the generated file, or add a validation step:

```bash
# Option A: Derive patterns from the auto-generated file
# In the CI step that generates the mobile test bundle, also generate the grep lists.

# Option B: Add a validation step after defining the patterns
ALL_FUNCTIONS=$(grep -oE 'function (run[A-Za-z0-9]+)' test/mobile/integration.auto.cjs | sed 's/function //')
MATCHED=$(echo "$ALL_FUNCTIONS" | grep -E "$IOS_GREP_A|$IOS_GREP_B" | sort -u)
UNMATCHED=$(comm -23 <(echo "$ALL_FUNCTIONS" | sort) <(echo "$MATCHED"))
if [ -n "$UNMATCHED" ]; then
  echo "::error::iOS test split is missing functions: $UNMATCHED"
  echo "Update IOS_GREP_A or IOS_GREP_B to include these."
  exit 1
fi
```

File: .github/workflows/integration-test-qvac-lib-infer-nmtcpp.yml [hunk 3]
```diff
@@ -222,12 +222,203 @@ jobs:
           echo "IndicTrans model files:"
           ls -la model/indictrans/
 
-      - name: Run integration test
+      - name: Run integration test (Linux/macOS)
+        if: ${{ matrix.platform != 'win32' }}
         working-directory: ${{ inputs.workdir || env.PKG_DIR }}
-        run: npm run test:integration ${{ matrix.platform }}-${{ matrix.arch }}
+        run: |
+          npm run test:integration ${{ matrix.platform }}-${{ matrix.arch }} 2>&1 | tee test-output.log
```
Comment (gianni-cor) [inline .github/workflows/integration-test-qvac-lib-infer-nmtcpp.yml:241]: **CodeQL flagged `pull_request_target` / artifact poisoning on this workflow — worth acknowledging.**

The GitHub Advanced Security bot flagged two findings on the lines you added here:
1. **Artifact poisoning** — `test-output.log` is created from `npm run test:integration` output and later uploaded as an artifact. Since this workflow accepts `workflow_call`, a parent workflow using `pull_request_target` could run untrusted PR code whose output flows into the artifact.
2. **Checkout of untrusted code** — same root cause; the workflow runs code from the calling ref.

These are arguably pre-existing (the workflow always ran `npm run test:integration`), but the new `tee` + artifact upload now persists output that wasn't persisted before.

**Review by gianni-cor — COMMENTED** [review]
Three suggestions for robustness — one surfaces an existing bug (dynamic-tools silently excluded from iOS mobile).

File: .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml (hunk 2 above)
Comment (gianni-cor) [reply .github/workflows/integration-mobile-test-qvac-lib-infer-llamacpp-llm.yml]: changing the workflow everytime we add a new test is not feasible.  so you can split the tests reading the names from  `integration.auto.cjs`

**Review by olyasir — APPROVED** [review]

**Review by Victor-Rodzko — APPROVED** [review]
